RFC Draft.md

```diff
@@ -382,9 +382,249 @@ const uploadedObjectReference = await uploadTask.result;
+  // only includes one query or mutation
+  const newTodo = res.data?.createTodo;
+
+  console.log('new todo created', newTodo);
+}
+
+function subscribeToCreate() {
+  const sub = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
+    graphqlOperation(onCreateTodo)
+  ).subscribe({
+    next: (message) => {
+      // once again, we could "sift up" the return value instead of providing it in two levels of depth
+      const newTodo = message.value?.data?.onCreateTodo;
+      
+      console.log('new todo created', newTodo);
+    }
+  });
+}
+```
+
+**Proposed Usage (v6)**
+```ts
+import { API, graphqlOperation } from 'aws-amplify';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  const res = await API.mutate(createTodo, {
+      input: createInput,
+  });
+
+  // The returned data is the result of the request. If there are more than one queries/mutations in a request,
+  // then the return value stays the same as v5. i.e. res.createTodo.data
+  const newTodo = res;
+
+  console.log('new todo created', newTodo);
+}
+
+function subscribeToCreate() {
+  const sub = API.subscribe(onCreateTodo).on({
+    next: (message) => {
+      // Return value immediately available instead of `value?.data?.onCreateTodo`
+      const newTodo = message;
+      console.log('new todo created', newTodo);
+    }
+  });
+}
+```
+
+## Type safety for GraphQL query, mutation, subscription inputs
+
+In v6, we want to ensure type safety on GraphQL inputs if you use one of the generated GraphQL query, mutation, or subscription.
+
+**Current Usage (v5)**
+```ts
+import { updateTodo } from './graphql/mutations';
+import { CreateTodoInput } from './API';
+
+const createInput: CreateTodoInput = {
+  name: todoName,
+  description,
+};
+
+const res = await API.graphql<GraphQLQuery<UpdateTodoMutation>>(
+  graphqlOperation(updateTodo, {
+    // passing an object of type CreateTodoInput (that's missing
+    // a required field for updates `id`) into an update mutation's input
+    // does not surface a type error. This will only throw a runtime error after
+    // the mutation request gets rejected by AppSync
+    input: createInput,
+  })
+);
+```
+
+**Proposed Usage (v6)**
+```ts
+import { updateTodo } from './graphql/mutations';
+import { CreateTodoInput } from './API';
+
+const createInput: CreateTodoInput = {
+  name: todoName,
+  description,
+};
+
+const res = await API.graphql<GraphQLQuery<UpdateTodoMutation>>(
+  graphqlOperation(updateTodo, {
+    input: createInput,
+    //^? Throws a type error. `input` must be of type `UpdateTodoInput`
```

Comment:
Not a great example, probably. If these types overlap sufficiently, TS will *not* throw an error here, despite the types appearing mismatched.

Reply:
I thought on update, you always have to provide an ID, where as on create it's optional.

Reply:
Oh yeah. 🤦 That makes sense!